File: ngmath/src/lib/gridpack/dsgrid/dsgrid2d.c
/************************************************************************
*                                                                       *
*                Copyright (C)  2000                                    *
*        University Corporation for Atmospheric Research                *
*                All Rights Reserved                                    *
*                                                                       *
*    The use of this Software is governed by a License Agreement.       *
*                                                                       *
************************************************************************/

#include <ncarg/ngmath.h>
#include "dstypes.h"
#include "dsproto.h"
#include "dsuhead.h"

/*
 *  Interpolate randomly-spaced 2D input data to a regularly spaced grid.
 */
double *c_dsgrid2d(int n, double x[], double y[], double u[],
                  int nx, int ny, double xo[], double yo[], int *ier)
/*
 *    Arguments
 *    ---------
 *        n - The number of input data points.
 *        x - An array of X coordinate values of the input data points.
 *        y - An array of Y coordinate values of the input data points.
 *        u - The functional value at coordinate (x,y).
 *       nx - The dimension of the array xo containing the X coordinate 
 *            values for the output grid.
 *       ny - The dimension of the array yo containing the Y coordinate 
 *            values for the output grid.
 *       xo - The array containing the X coordinate values for the output 
 *            grid (must be monotone increasing, but need not be equally 
 *            spaced.
 *       yo - The array containing the Y coordinate values for the output 
 *            grid (must be monotone increasing, but need not be equally 
 *            spaced.
 *     *ier - An error return value ( = 0 is no error).
 *
 *   Return value
 *   ------------
 *      A pointer to the first element of a linear array that is
 *      laid out as a 2D array (i.e. the last subscript varies fastest)
 *      of dimension: nx by ny.
 */
{
  
  static double perror = 1.;
  double zo[1] = {0.}, *z, *retval;
  
  z = (double *) calloc(n, sizeof(double));
  if (z == NULL) {
    DSErrorHnd(13, "c_dsgrid2d", stderr, "\n");
    *ier = ds_error_status;
    return(&perror);
  }

  retval = c_dsgrid3d(n, x, y, z, u, nx, ny, 1, xo, yo, zo, ier);

  free(z);
  return(retval);
}

void c_dspnt2d(int n, double xi[], double yi[], double zi[],
             int m, double xo[], double yo[], double zo[], int *ier)
{
  int    i;
  double xt[1], yt[1];

  for (i = 0; i < m; i++) {
    xt[0] = xo[i];
    yt[0] = yo[i];
    zo[i] = *c_dsgrid2d(n, xi, yi, zi, 1, 1, xt, yt, ier);
  }
}

File: ngmath/man/dsgrid/dsgrid2d.m
.\"
.\"     $Id: dsgrid2d.m,v 1.5 2008-07-27 03:35:36 haley Exp $
.\"
.TH DSGRID2D 3NCARG "September 1997-1998" UNIX "NCAR GRAPHICS"
.SH NAME
DSGRID2D - double precision Fortran entry for iterpolation from 
2D random data to an output grid.
.SH SYNOPSIS
CALL DSGRID2D (NPNTS, X, Y, Z, NUMXOUT, NUMYOUT, XO, YO, ZO, IER)
.SH DESCRIPTION
.IP NPNTS 12
(Integer, Input) - The number of input data points.
.IP X 12
(Double precision, Input) - An array containing the 
X coordinates of the input data points.
.IP Y 12
(Double precision, Input) - An array containing the Y 
coordinates of the input data points.
.IP Z 12
(Double precision, Input) - An array containing the functional values 
of the input data points. That is, Z(L) is the value of the input function at 
coordinate (X(L),Y(L)), for L=1,NPNTS. 
.IP NUMXOUT 12
(Integer, Input) - The number of X values in the output grid.
.IP NUMYOUT 12
(Integer, Input) - The number of Y values in the output grid.
.IP XO 12
(Double precision, Input) - An array of dimension NUMXOUT containing the X 
coordinates of the output data grid. The values in XO must be 
increasing, but need not be equally spaced. 
.IP YO 12
(Double precision, Input) - An array of dimension NUMYOUT containing the Y 
coordinates of the output data grid. The values in YO must be 
increasing, but need not be equally spaced. 
.IP ZO 12
(Double precision, Output) -  A two-dimensional array of
dimension NUMXOUT x NUMYOUT
containing the interpolated functional values. ZO(I,J) is the interpolated
value at grid point (XO(I),YO(J)).
.IP IER 12
(Integer, Output) - An error return value. If IER is returned as 0, then
no errors were detected. If IER is non-zero, then refer to the man
page for dsgrid_errors for details.
.SH USAGE
DSGRID2D is the double precision gridding function for 2D data 
in the Dsgrid package.
The behavior of DSGRID2D is controlled by values set for the various
parameters described in dsgrid_params.  Values for these parameters
can be set using the subroutines DSSETI, DSSETRD, and DSSETC.
.sp
In almost all cases the ngmath function NATGRIDD is superior to
DSGRID2D for 2D interpolation.  See the man page for NATGRIDD for details.
.SH ACCESS
To use DSGRID2D, load the NCAR Graphics library ngmath.
.SH SEE ALSO
dsgrid,
dsgrid_params, 
dsseti, 
dsgeti, 
dssetrd, 
dsgetrd, 
dssetc, 
dsgetc.
.sp
Complete documentation for Dsgrid is available at URL
.br
http://ngwww.ucar.edu/ngdoc/ng/ngmath/dsgrid/dshome.html
.SH COPYRIGHT
Copyright (C) 2000
.br
University Corporation for Atmospheric Research
.br

The use of this Software is governed by a License Agreement.
